mqtt/mqtt-bridge/src/pump.rs

```diff
@@ -0,0 +1,347 @@
+        let mut loader = self.loader.clone();
+        let mut client_shutdown = self.client_shutdown.clone();
+        let pump_context = self.pump_context.clone();
+
+        // egress pump
+        let egress_pump = async {
+            let mut receive_fut = loader_shutdown_rx.into_stream();
+
+            info!("{} starting egress publication processing", pump_context);
+
+            loop {
+                let mut publish_handle = publish_handle.clone();
+                match select(receive_fut.next(), loader.try_next()).await {
+                    Either::Left((shutdown, _)) => {
+                        info!(
+                            "{} received shutdown signal for egress messages",
+                            pump_context
+                        );
+                        if shutdown.is_none() {
+                            error!("{} has unexpected behavior from shutdown signal while signalling bridge pump shutdown", pump_context);
+                        }
+
+                        break;
+                    }
+                    Either::Right((loaded_element, _)) => {
+                        debug!("{} extracted publication from store", pump_context);
+
+                        if let Ok(Some((key, publication))) = loaded_element {
+                            debug!("{} publishing {:?}", pump_context, key);
+                            if let Err(e) = publish_handle.publish(publication).await {
+                                let err_msg = format!("{} failed publish", pump_context);
+                                error!(message = err_msg.as_str(), err = %e);
+                            }
+
+                            if let Err(e) = persist.remove(key) {
+                                let err_msg = format!(
+                                    "{} failed removing publication from store",
+                                    pump_context
+                                );
+                                error!(message = err_msg.as_str(), err = %e);
+                            }
+                        }
+                    }
+                }
+            }
+
+            info!("{} stopped sending egress messages", pump_context);
+        };
+
+        // ingress pump
+        let pump_context = self.pump_context.clone();
+        let ingress_pump_fut = async {
+            debug!("{} starting ingress publication processing", pump_context);
+            self.client.handle_events().await;
+        };
+
+        let egress_pump = egress_pump.fuse();
+        let ingress_pump = ingress_pump_fut.fuse();
+        let mut shutdown = shutdown.fuse();
+        pin_mut!(egress_pump, ingress_pump);
+
+        select! {
+            _ = egress_pump => {
+                error!("{} failed egress publication loop and exited", pump_context);
+            },
+            _ = ingress_pump => {
+                error!("{} failed ingress publication loop and exited", pump_context);
+            },
```

Comment:
We need to shutdown and exit in both of the above cases. Currently it won't work.